# CH-4. LOOPS

## 4.1 While Loop

 
The while loop is like a repeated if statement. The code is executed over and over again, as long as the condition is True. Have another look at its recipe.

        while condition :
            expression
    

WL is Numerically calculating model

"repeating action until condition is met”

Example:

    ● Error starts at 50
    ● Divide error by 4 on every run
    ● Continue until error no longer > 1

**EXERCISE-1: Basic while loop**

Create the variable offset with an initial value of 8.

Code a while loop that keeps running as long as offset is not equal to 0. 

Inside the while loop:
Print out the sentence "correcting...".

Next, decrease the value of offset by 1. You can do this with offset = offset - 1.
Finally, still within your loop, print out offset so you can see how it changes.

In [1]:
# Initialize offset
offset = 8

# Code the while loop
while offset != 0:
    print("correcting...")
    offset = offset-1
    print(offset)

correcting...
7
correcting...
6
correcting...
5
correcting...
4
correcting...
3
correcting...
2
correcting...
1
correcting...
0


**EXERCISE-2: While Loop + Conditionals**

Inside the while loop, complete the if-else statement:

If offset is greater than zero, you should decrease offset by 1.

Else, you should increase offset by 1.

If you've coded things correctly, hitting Submit Answer should work this time.

In [2]:
# Initialize offset
offset = -6

# Code the while loop
while offset != 0 :
    print("correcting...")
    if offset>0 :
        offset = offset-1
    else : 
        offset = offset +1
    print(offset)

correcting...
-5
correcting...
-4
correcting...
-3
correcting...
-2
correcting...
-1
correcting...
0


## 4.2 for loops

### 4.2.1 Loop over a list

**Exercise-3: for Loops**

Write a for loop that iterates over all elements of the areas list and prints out every element separately.

In [3]:
# areas list
areas = [11.25, 18.0, 20.0, 10.75, 9.50]

# Code the for loop
for x in areas:
    print(x)

11.25
18.0
20.0
10.75
9.5


### 4.2.2 Loop with Indexes and values

Using a for loop to iterate over a list only gives you access to every list element in each run, one after the other. 

If you also want to access the index information, so where the list element you're iterating over is located, you can use `enumerate()`.

`enumerate()` produces two values:
    
    i. the index: placed on the first variable on the for loop
    ii. and the value: placed on the second value on the variable list of for loop

**Exercise-4: Loop with index**

Adapt the for loop in the sample code to use enumerate() and use two iterator variables.

Update the print() statement so that on each run, a line of the form "room x: y" should be printed, where x is the index of the list element and y is the actual list element, i.e. the area. 

Make sure to print out this exact string, with the correct spacing.

In [5]:
# areas list
areas = [11.25, 18.0, 20.0, 10.75, 9.50]

# Change for loop to use enumerate() and update print()
for index, value in enumerate(areas) :
    print("room"+ str(index)+":"+str(value))

room0:11.25
room1:18.0
room2:20.0
room3:10.75
room4:9.5


Adapt the print() function in the for loop on the right so that the first printout becomes 

"room 1: 11.25", the second one "room 2: 18.0" and so on.

In [6]:
# areas list
areas = [11.25, 18.0, 20.0, 10.75, 9.50]

# Code the for loop
for index, area in enumerate(areas) :
    print("room " + str(index+1) + ": " + str(area))

room 1: 11.25
room 2: 18.0
room 3: 20.0
room 4: 10.75
room 5: 9.5


### 4.2.3 Loop over list of lists

Remember the house variable from the Intro to Python course? 

Have a look at its definition on the right. It's basically a list of lists, where each sublist contains the name and area of a room in your house.

It's up to you to build a for loop from scratch this time!

**Exercise-5: Loop over list of lists**

Write a for loop that goes through each sublist of house and prints out the x is y sqm, where x is the name of the room and y is the area of the room.

In [50]:
# house list of lists
house = [["hallway", 11.25], 
         ["kitchen", 18.0], 
         ["living room", 20.0], 
         ["bedroom", 10.75], 
         ["bathroom", 9.50]]
         
print("the "+house[1][0]+" is "+ str(house[1][1])+" sqm") 
print("the "+house[2][0]+" is "+ str(house[2][1])+" sqm")
print("the "+house[3][0]+" is "+ str(house[3][1])+" sqm")
print("the "+house[4][0]+" is "+ str(house[4][1])+" sqm")

the kitchen is 18.0 sqm
the living room is 20.0 sqm
the bedroom is 10.75 sqm
the bathroom is 9.5 sqm


In [51]:
for x in house:
    print("the " + x[0] + " is " + str(x[1]) + " sqm")

the hallway is 11.25 sqm
the kitchen is 18.0 sqm
the living room is 20.0 sqm
the bedroom is 10.75 sqm
the bathroom is 9.5 sqm




We have already seen how loop works on strings and list. 

**The question is how about other data structures?**

We can use the same forloop construct but the way how we define the sequence depends on the type of data structure at hand.

Looping depends on the data structure (dictionary, or numpy array) at hand. 


To generate items we use `.items()` method for `dictionaries`.

To generate items we use `np.nditer()` function for `numpyArrays`.

### 4.2.4 Loop over dictionary

we use `.items()` method

In Python 3, we need the `items()` `method` to loop over a `dictionary`.

In [56]:
world = { "afghanistan":30.55, 
          "albania":2.77,
          "algeria":39.21 }

for key, value in world.items() :
    print(key + " -- " + str(value))

afghanistan -- 30.55
albania -- 2.77
algeria -- 39.21


**Exercise-6: LOOP OVER A DICTIONARY"**

Write a for loop that goes through each key:value pair of europe. 

On each iteration, "the capital of x is y" should be printed out, where x is the key and y is the value of the pair.

In [61]:
# Definition of dictionary
europe = {'Germany':'Berlin', 'France':'Paris', 'Sweden':'Stockholm',
          'Norway':'Oslo' }
          
# Iterate over europe
for key, value in europe.items():
    print("the capital of "+ key +"is: "+ value)

the capital of Germanyis: Berlin
the capital of Franceis: Paris
the capital of Swedenis: Stockholm
the capital of Norwayis: Oslo


### 4.2.5 Loop over Numpy array

 **For a 1D Numpy array:** - looping over all elements can be as simple as:

    for x in my_array :
        ...
        
**For a 2D Numpy array:**  it is a bit `more complicated`. 

A 2D array is built up of multiple 1D arrays. 

To explicitly iterate over all separate elements of a `multi-dimensional array`, you'll need this syntax:

    for x in np.nditer(my_array) :
        ...
        

**EXERCISE -7: LOOP OVER 2D ARRAY**

Import the numpy package under the local alias np.

1. Write a for loop that iterates over all elements in `np_height` and prints out `"x inches"` for each element, where `x is the value` in the array.

2. Write a for loop that visits every element of the **meas array** and prints it out.

In [63]:
# Import numpy as np
import numpy as np
np_height = [215, 195, 200, 215, 229, 240]
# For loop over np_height
for x in np_height:
    print(str(x)+" inches")

215 inches
195 inches
200 inches
215 inches
229 inches
240 inches


In [66]:
import numpy as np
np_height = np.array([1.73, 1.68, 1.71, 1.89, 1.79])
np_weight = np.array([65.4, 59.2, 63.6, 88.4, 68.7])
meas = np.array([np_height, np_weight])
print(meas)

[[ 1.73  1.68  1.71  1.89  1.79]
 [65.4  59.2  63.6  88.4  68.7 ]]


In [68]:
# loop that visits every element of the **meas array** and prints it out.

for val in np.nditer(meas):
    print(val)

1.73
1.68
1.71
1.89
1.79
65.4
59.2
63.6
88.4
68.7


### Recap

#### Dictionary
● for key, val in my_dict.items() 

#### Numpy array
● for val in np.nditer(my_array) :

### 4.2.6 Loop on Pandas Data frame

In [70]:
import pandas as pd
brics = pd.read_csv("brics.csv", index_col =0)
print(brics)

         country    capital    area  population
BR        Brazil   Brasilia   8.516      200.40
RU        Russia     Moscow  17.100      143.50
IN         India  New Delhi   3.286     1252.00
CH         China    Beijing   9.597     1357.00
SA  South Africa   Pretoria   1.221       52.98


**I. USING BASIC  FOR LOOP**

Simply returns the `column names`

In [72]:
for x in brics:
    print(x)

country
capital
area
population


Well, that's not bad. 

But what `we want is to iterate over the rows` most of the time.

How do we do that??


**II. ITERROWS METHOD**

iterrows() method generates two pices of data for each iteration. 

1. The `lebel` of the `row` and 
    
2. The `actual data` in the row as a `panda series`
    

In [84]:
import pandas as pd
brics = pd.read_csv("brics.csv", index_col =0)

for label, row in brics.iterrows():
    print(label)

BR
RU
IN
CH
SA


In [85]:
for label, row in brics.iterrows():
    print(row, "\n")

country         Brazil
capital       Brasilia
area             8.516
population       200.4
Name: BR, dtype: object 

country       Russia
capital       Moscow
area            17.1
population     143.5
Name: RU, dtype: object 

country           India
capital       New Delhi
area              3.286
population         1252
Name: IN, dtype: object 

country         China
capital       Beijing
area            9.597
population       1357
Name: CH, dtype: object 

country       South Africa
capital           Pretoria
area                 1.221
population           52.98
Name: SA, dtype: object 



In [112]:
import pandas as pd
brics = pd.read_csv("brics.csv", index_col =0)

for label, row in brics.iterrows():
    print(label)
    print(row, "\n")

BR
country         Brazil
capital       Brasilia
area             8.516
population       200.4
Name: BR, dtype: object 

RU
country       Russia
capital       Moscow
area            17.1
population     143.5
Name: RU, dtype: object 

IN
country           India
capital       New Delhi
area              3.286
population         1252
Name: IN, dtype: object 

CH
country         China
capital       Beijing
area            9.597
population       1357
Name: CH, dtype: object 

SA
country       South Africa
capital           Pretoria
area                 1.221
population           52.98
Name: SA, dtype: object 



In [126]:
for label, row in cars.iterrows():
    print(label, "\n")
    print(row)

US 

cars_per_cap              809
country         United States
drives_right             True
Name: US, dtype: object
AUS 

cars_per_cap           31
country         Australia
drives_right        False
Name: AUS, dtype: object
JPN 

cars_per_cap      588
country         Japan
drives_right    False
Name: JPN, dtype: object
IN 

cars_per_cap       18
country         India
drives_right    False
Name: IN, dtype: object
RU 

cars_per_cap       200
country         Russia
drives_right      True
Name: RU, dtype: object
MOR 

cars_per_cap         70
country         Morocco
drives_right       True
Name: MOR, dtype: object
EG 

cars_per_cap       45
country         Egypt
drives_right     True
Name: EG, dtype: object


#### Selective Print

In [127]:
import pandas as pd
brics = pd.read_csv("brics.csv", index_col =0)
print(brics)

         country    capital    area  population
BR        Brazil   Brasilia   8.516      200.40
RU        Russia     Moscow  17.100      143.50
IN         India  New Delhi   3.286     1252.00
CH         China    Beijing   9.597     1357.00
SA  South Africa   Pretoria   1.221       52.98


In [128]:
#selective print

for label, row in brics.iterrows():
    print(label,":",row["capital"])

BR : Brasilia
RU : Moscow
IN : New Delhi
CH : Beijing
SA : Pretoria


In [129]:
#selective print

for label, row in brics.iterrows():
    print(label,":",row["country"],":",row["capital"] )

BR : Brazil : Brasilia
RU : Russia : Moscow
IN : India : New Delhi
CH : China : Beijing
SA : South Africa : Pretoria


In [130]:
#selective print

for label, row in brics.iterrows():
    print(label,":",row["country"],":",row["capital"] )

BR : Brazil : Brasilia
RU : Russia : Moscow
IN : India : New Delhi
CH : China : Beijing
SA : South Africa : Pretoria


### ADVANCED STUFF: using loop to modify our data frame

Let's say we wanted **to add a column to our data frame** that indicates the number of characters for a country in a  country's column. 

`How do we add one more column in our data frame using loop?`

In [153]:
import pandas as pd
brics = pd.read_csv("brics.csv", index_col =0)

for label, row in brics.iterrows():
    brics.loc[label, "name_length"] = len(row["country"]) #creating a series on every iteration
    print(brics)
    print()


         country    capital    area  population  name_length
BR        Brazil   Brasilia   8.516      200.40          6.0
RU        Russia     Moscow  17.100      143.50          NaN
IN         India  New Delhi   3.286     1252.00          NaN
CH         China    Beijing   9.597     1357.00          NaN
SA  South Africa   Pretoria   1.221       52.98          NaN

         country    capital    area  population  name_length
BR        Brazil   Brasilia   8.516      200.40          6.0
RU        Russia     Moscow  17.100      143.50          6.0
IN         India  New Delhi   3.286     1252.00          NaN
CH         China    Beijing   9.597     1357.00          NaN
SA  South Africa   Pretoria   1.221       52.98          NaN

         country    capital    area  population  name_length
BR        Brazil   Brasilia   8.516      200.40          6.0
RU        Russia     Moscow  17.100      143.50          6.0
IN         India  New Delhi   3.286     1252.00          5.0
CH         China    Be

## using `apply()` Function

The above method is slow and computationally intensive for a big data. 

So we can do this using with out iteration using the apply() function.

**WAY MORE EFFICIENT** 

In [132]:
import pandas as pd
brics = pd.read_csv("brics.csv", index_col = 0)
brics["name_length"] = brics["country"].apply(len)
print(brics)

         country    capital    area  population  name_length
BR        Brazil   Brasilia   8.516      200.40            6
RU        Russia     Moscow  17.100      143.50            6
IN         India  New Delhi   3.286     1252.00            5
CH         China    Beijing   9.597     1357.00            5
SA  South Africa   Pretoria   1.221       52.98           12


**EXERCISE-8: Loop over DataFrame**

Write a for loop that iterates over the `rows of cars` and on each iteration perform two print() calls: 

one to print out the row label and 

one to print out all of the rows contents.

In [122]:
import pandas as pd
cars = pd.read_csv("cars.csv", index_col =0)
print(cars)

     cars_per_cap        country  drives_right
US            809  United States          True
AUS            31      Australia         False
JPN           588          Japan         False
IN             18          India         False
RU            200         Russia          True
MOR            70        Morocco          True
EG             45          Egypt          True


In [123]:
for label, row in cars.iterrows():
    print(label, "\n")
    print(row)

US 

cars_per_cap              809
country         United States
drives_right             True
Name: US, dtype: object
AUS 

cars_per_cap           31
country         Australia
drives_right        False
Name: AUS, dtype: object
JPN 

cars_per_cap      588
country         Japan
drives_right    False
Name: JPN, dtype: object
IN 

cars_per_cap       18
country         India
drives_right    False
Name: IN, dtype: object
RU 

cars_per_cap       200
country         Russia
drives_right      True
Name: RU, dtype: object
MOR 

cars_per_cap         70
country         Morocco
drives_right       True
Name: MOR, dtype: object
EG 

cars_per_cap       45
country         Egypt
drives_right     True
Name: EG, dtype: object


**EXERCISE-9: Loop over DataFrame: Selective Print**

Using the iterators lab and row, adapt the code in the for loop such that:

    the first iteration prints out "US: 809", 
    the second iteration "AUS: 731", 
    and so on.


The output should be in the form

    "country: cars_per_cap". 

Make sure to print out this exact string (with the correct spacing).

You can use `str()` to convert your integer data to a string so that you can print it in conjunction with the `country label`.

In [133]:
import pandas as pd
cars = pd.read_csv("cars.csv", index_col =0)
print(cars)

     cars_per_cap        country  drives_right
US            809  United States          True
AUS            31      Australia         False
JPN           588          Japan         False
IN             18          India         False
RU            200         Russia          True
MOR            70        Morocco          True
EG             45          Egypt          True


In [138]:
for label, row in cars.iterrows():
    print(label+":"+str(row["cars_per_cap"]))

US:809
AUS:31
JPN:588
IN:18
RU:200
MOR:70
EG:45


**EXERCISE-10: Loop over DataFrame: Add a new Column**

Use a for loop `to add a new column`,

    named COUNTRY, that contains a uppercase version of the country names in the "country" col. 
    You can use the string method upper() for this.
    
To see if your code worked, print out cars. Don't indent this code, so that it's not part of the for loop.

In [143]:
import pandas as pd
cars = pd.read_csv("cars.csv", index_col =0)
print(cars)

     cars_per_cap        country  drives_right
US            809  United States          True
AUS            31      Australia         False
JPN           588          Japan         False
IN             18          India         False
RU            200         Russia          True
MOR            70        Morocco          True
EG             45          Egypt          True


In [156]:
for label, row in cars.iterrows():
    cars.loc[label, "COUNTRY"] = row["country"].upper()#creating a series on every iteration
print(cars)

     cars_per_cap        country  drives_right        COUNTRY
US            809  United States          True  UNITED STATES
AUS            31      Australia         False      AUSTRALIA
JPN           588          Japan         False          JAPAN
IN             18          India         False          INDIA
RU            200         Russia          True         RUSSIA
MOR            70        Morocco          True        MOROCCO
EG             45          Egypt          True          EGYPT


In [159]:
import pandas as pd
cars = pd.read_csv("cars.csv", index_col = 0)
cars["COUNTRY"] = cars["country"].apply(str.upper)
print(cars)

     cars_per_cap        country  drives_right        COUNTRY
US            809  United States          True  UNITED STATES
AUS            31      Australia         False      AUSTRALIA
JPN           588          Japan         False          JAPAN
IN             18          India         False          INDIA
RU            200         Russia          True         RUSSIA
MOR            70        Morocco          True        MOROCCO
EG             45          Egypt          True          EGYPT
